presto-main/src/main/java/com/facebook/presto/sql/gen/InCodeGenerator.java

```diff
@@ -102,25 +119,41 @@ public ByteCodeNode generateExpression(Signature signature, ByteCodeGeneratorCon
         if (constantValues.size() < 1000) {
             ByteCodeBlock switchCaseBlocks = new ByteCodeBlock();
             LookupSwitch.LookupSwitchBuilder switchBuilder = lookupSwitchBuilder();
-            for (Map.Entry<Integer, Collection<ByteCodeNode>> bucket : hashBuckets.asMap().entrySet()) {
-                LabelNode label = new LabelNode("inHash" + bucket.getKey());
-                switchBuilder.addCase(bucket.getKey(), label);
-                Collection<ByteCodeNode> testValues = bucket.getValue();
-
-                ByteCodeBlock caseBlock = buildInCase(generatorContext, scope, type, label, match, defaultLabel, testValues, false);
-                switchCaseBlocks
-                        .append(caseBlock.setDescription("case " + bucket.getKey()));
+            if (areSmallPrimitives) {
+                for (Object object : constantValues) {
+                    int constVal = ((Long) object).intValue();
+                    LabelNode label = new LabelNode("inVal" + constVal);
+                    switchBuilder.addCase(constVal, label);
+
+                    ByteCodeBlock caseBlock = new ByteCodeBlock().visitLabel(label);
+                    caseBlock.gotoLabel(match);
+                    switchCaseBlocks.append(caseBlock.setDescription("case " + constVal));
```

Comment:
Instead of 2-step jump (to `label` first, and then to `match`), I believe you can jump directly to `match`. Give it a try.